### PR TITLE
Develop table to view level

### PR DESCRIPTION
**Pull request by PedroMagalhaes20**

### CODE REVIEW

File: backend/src/main/java/saci/application/LevelController.java [hunk 1]
```diff
@@ -1,62 +1,64 @@
...
     public ResponseEntity<Void> deleteLevelById(@PathVariable long levelId) {
         try {
-            LevelService.deleteLevelById(levelId);
+            levelService.deleteLevelById(levelId);
             return ResponseEntity.ok().build();
         } catch (NotFoundException e) {
             return ResponseEntity.notFound().build();
         }
     }
-
```
Comment (vitorlofonseca) [inline backend/src/main/java/saci/application/LevelController.java:44]: Please, don't take out already developed features

File: backend/src/main/java/saci/application/LevelController.java [hunk 2]
```diff
@@ -1,62 +1,64 @@
...
+        }
+        return levels.isEmpty() ? ResponseEntity.noContent().build() : ResponseEntity.ok(levels);
+    }
 
     @Operation(summary = "Delete Level by ID", description = "Deletes a Level based on its ID")
     @ApiResponses(
             value = {
-                @ApiResponse(responseCode = "200", description = "Level deleted successfully"),
-                @ApiResponse(responseCode = "404", description = "Level not found")
+                    @ApiResponse(responseCode = "200", description = "Level deleted successfully"),
```
Comment (vitorlofonseca) [inline backend/src/main/java/saci/application/LevelController.java:52]: Please, run `mvn spotless:apply` to take out the blank spaces

File: backend/src/main/java/saci/domain/model/Level.java [hunk 3]
```diff
@@ -18,7 +17,6 @@ public class Level {
     private Long id;
 
     @Column(name = "name")
-    @NotBlank(message = "Name is mandatory")
```
Comment (vitorlofonseca) [inline backend/src/main/java/saci/domain/model/Level.java:21]: Why take out these validations?

File: backend/src/main/java/saci/application/LevelController.java (hunk 2 above)
Comment (PedroMagalhaes20) [reply backend/src/main/java/saci/application/LevelController.java:52]: done

File: backend/src/main/java/saci/domain/service/LevelService.java [hunk 4]
```diff
@@ -25,19 +36,11 @@ public void deleteLevelById(long levelId) {
         }
     }
 
-    public Level createLevel(Level level) {
```
Comment (vitorlofonseca) [inline backend/src/main/java/saci/domain/service/LevelService.java:28]: Please, come back with this method

File: frontend/package.json [hunk 5]
```diff
@@ -17,7 +17,8 @@
     "eslint-plugin-playwright": "^1.6.0",
     "pinia": "^2.1.7",
     "vue": "^3.4.21",
-    "vue-router": "^4.3.0"
+    "vue-router": "^4.3.0",
+    "vuedraggable": "^4.1.0"
```
Comment (vitorlofonseca) [inline frontend/package.json:21]: Remove this dependency

File: frontend/src/stores/levels/index.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import { defineStore } from 'pinia'
+import { getLevels } from './getters'
+import { levels } from './state'
+
+export const useLevelsStore = defineStore('levels', () => {
+  return {
+    levels,
+    getLevels,
+    async fetchLevelsByRoleId(roleId: string) {
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/index.ts:9]: NEVER do this. This is a getter, and should be in the getters file

File: frontend/src/stores/levels/index.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+      try {
+        const response = await fetch(`http://localhost:8080/api/levels?role_id=${roleId}`)
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/index.ts:11]: Don't use fetch, use the methods in the http.ts file, that is already wrapping the fetch

File: frontend/src/views/LevelView.vue [hunk 8]
```diff
@@ -65,26 +65,28 @@ const redirectToRoleView = () => {
...
       return
     }
     if (newLevel.value.maxCoefficient <= newLevel.value.minCoefficient) {
       ErrorMessage('MaxRange should be bigger than MinRange')
       return
     }
-    await roleStore.addLevel(newLevel.value, role.value!)
+
     redirectToRoleView()
+    await roleStore.addLevel(newLevel.value, role.value!)
```
Comment (vitorlofonseca) [inline frontend/src/views/LevelView.vue:82]: Come back with the previous version, please

File: frontend/src/views/Viewlevelstable.vue
Comment (vitorlofonseca) [inline frontend/src/views/Viewlevelstable.vue:1]: "ViewLevelStable" or "ViewLevelsTable"? CapitalCase, please

File: frontend/src/views/Viewlevelstable.vue [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+const redirectToLevelView = () => {
+  router.push('/level-view')
+}
+
+onMounted(() => {
+  fetchLevelsByRoleId()
+})
+
+watch(
```
Comment (vitorlofonseca) [inline frontend/src/views/Viewlevelstable.vue:43]: Do you need this watch?

File: frontend/src/views/Viewlevelstable.vue [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import { ref, onMounted, watch } from 'vue'
+import { ElTable, ElTableColumn, ElButton } from 'element-plus'
+import { useRouter, useRoute } from 'vue-router'
+import { getLevels } from '@/stores/levels/getters'
+import { useLevelsStore } from '@/stores/levels'
+
+const router = useRouter()
+const route = useRoute()
+
+const roleId = ref(route.params.roleId?.toString() || '')
```
Comment (vitorlofonseca) [inline frontend/src/views/Viewlevelstable.vue:26]: I don't think you need this ref. If you get this value inside the fetch function, is already enough

File: frontend/src/views/Viewlevelstable.vue [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+const levels = getLevels
+
+const fetchLevelsByRoleId = async () => {
+  const levelsStore = useLevelsStore()
```
Comment (vitorlofonseca) [inline frontend/src/views/Viewlevelstable.vue:31]: ANY store should be accessible at the component level, not at a function level

File: backend/src/main/java/saci/infrastructure/LevelRepository.java [hunk 12]
```diff
@@ -10,8 +12,10 @@
 public interface LevelRepository extends JpaRepository<Level, Long> {
     Optional<Level> findByName(String name);
 
+    List<Level> findByRoleId(Long roleId);
+
     @Query(
-            "SELECT COUNT(*) FROM Level l WHERE l.roleId = :roleId AND (l.minCoefficient < :maxCoefficient AND l.maxCoefficient > :minCoefficient)")
+            "SELECT COUNT(l) FROM Level l WHERE l.roleId = :roleId AND (l.minCoefficient < :maxCoefficient AND l.maxCoefficient > :minCoefficient)")
```
Comment (vitorlofonseca) [inline backend/src/main/java/saci/infrastructure/LevelRepository.java:18]: Why this change?

File: frontend/src/router/index.ts [hunk 13]
```diff
@@ -31,6 +31,13 @@ const router = createRouter({
       path: '/level-view',
       component: LevelView
     },
+    {
+      path: '/level-table/:roleId',
+      component: ViewLevelsTable,
+      props: (route) => ({
+        roleId: Array.isArray(route.params.roleId) ? route.params.roleId[0] : route.params.roleId
```
Comment (vitorlofonseca) [inline frontend/src/router/index.ts:38]: In what context the roleId would be an array?

File: frontend/src/stores/levels/actions.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import { levels } from './state'
+import { get } from '@/services/http'
+
+export async function fetchLevelsByRoleId(roleId: string) {
+  try {
+    const data: any[] = await get(`/levels?role_id=${roleId}`) // Assuming data is an array of any type
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:6]: Your code should be the comment. Can you get rid of comments?

File: backend/src/main/java/saci/domain/service/LevelService.java [hunk 15]
```diff
@@ -34,10 +26,33 @@ public Level createLevel(Level level) {
...
+    public void deleteLevelById(long levelId) {
+        Optional<Level> level = levelRepository.findById(levelId);
+        if (level.isPresent()) {
+            levelRepository.deleteById(levelId);
+        } else {
+            throw new NotFoundException("Level not found with ID: " + levelId);
+        }
+    }
+
+    public Optional<Level> findById(long levelId) {
```
Comment (vascolusitan) [inline backend/src/main/java/saci/domain/service/LevelService.java:51]: Are yo using this method anywhere?

File: backend/src/main/java/saci/application/LevelController.java [hunk 16]
```diff
@@ -1,30 +1,51 @@
...
 import io.swagger.v3.oas.annotations.responses.ApiResponses;
+import java.util.List;
 import lombok.RequiredArgsConstructor;
+import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
-import org.springframework.web.bind.annotation.DeleteMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (vascolusitan) [inline backend/src/main/java/saci/application/LevelController.java:13]: Star import.

File: backend/src/main/java/saci/application/LevelController.java [hunk 17]
```diff
@@ -1,30 +1,51 @@
...
+                                                    schema = @Schema(implementation = Level.class)))
+                        })
+            })
+    @GetMapping
+    public ResponseEntity<List<Level>> getLevels(@RequestParam(required = false) Long role_id) {
+        List<Level> levels;
+        if (role_id != null) {
+            levels = levelService.getLevelsByRoleId(role_id);
+        } else {
+            levels = levelService.getLevels();
```
Comment (vascolusitan) [inline backend/src/main/java/saci/application/LevelController.java:45]: In any of the frontends are you calling this endpoint without roleId? 
If not you don't need this logic (same in the service).

File: backend/src/main/java/saci/application/LevelController.java [hunk 18]
```diff
@@ -1,30 +1,51 @@
...
+                        content = {
+                            @Content(
+                                    mediaType = "application/json",
+                                    array =
+                                            @ArraySchema(
+                                                    schema = @Schema(implementation = Level.class)))
+                        })
+            })
+    @GetMapping
+    public ResponseEntity<List<Level>> getLevels(@RequestParam(required = false) Long role_id) {
```
Comment (vascolusitan) [inline backend/src/main/java/saci/application/LevelController.java:40]: Use camel case for java variables.

File: frontend/src/stores/levels/actions.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
+import { levels } from './state'
+import { get } from '@/services/http'
+
+export async function fetchLevelsByRoleId(roleId: string) {
+  try {
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:5]: Try-catches should be in the component (Views), not in the actions. Please, put your error treatment in the View

File: backend/src/main/java/saci/domain/service/LevelService.java [hunk 20]
```diff
@@ -34,10 +26,33 @@ public Level createLevel(Level level) {
...
+
         Optional<Level> optionalLevel = levelRepository.findByName(level.getName());
         if (optionalLevel.isPresent()) {
             throw new AlreadyExistsException("Level name already exists");
         }
+
         return levelRepository.save(level);
     }
+
+    public List<Level> getLevels() {
```
Comment (fromgonie) [inline backend/src/main/java/saci/domain/service/LevelService.java:38]: A scenario where we need to retrieve all levels, without filter based on role doesn`t exist. Please, get rid of this method

File: frontend/src/services/http.ts [hunk 21]
```diff
@@ -19,7 +19,11 @@ const fetchWrapper = async (url: string, init: RequestInit): Promise<Response> =
   })
 }
 
-const doRequest = async <T>(url: string, method: string, body?: object | string): Promise<T> => {
+const doRequest = async <T>(
+  url: string,
+  method: string,
+  body?: object | string
+): Promise<T | string> => {
```
Comment (fromgonie) [inline frontend/src/services/http.ts:26]: Why?

File: frontend/src/stores/levels/actions.ts [hunk 22]
```diff
@@ -0,0 +1,17 @@
+import { levels } from './state'
+import { get } from '@/services/http'
+
+export async function fetchLevelsByRoleId(roleId: string) {
+  const data: any[] | string = await get(`/levels?role_id=${roleId}`)
```
Comment (fromgonie) [inline frontend/src/stores/levels/actions.ts:5]: Never use any type. You are retrieving levels of a given role, right? So, what`s the type of your return?

File: frontend/src/stores/levels/actions.ts [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+  if (Array.isArray(data)) {
```
Comment (fromgonie) [inline frontend/src/stores/levels/actions.ts:6]: You don`t need this if statement, if you type the variable correctly

File: frontend/src/stores/levels/actions.ts [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+    levels.value = data.map((level: any) => ({
```
Comment (fromgonie) [inline frontend/src/stores/levels/actions.ts:7]: Please, put these levels in the role state. Each role has its levels, so you can update there, and get rid of levels state file

File: frontend/src/stores/levels/actions.ts [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+      id: level.id,
+      name: level.name,
+      minCoefficient: 0,
+      maxCoefficient: 0,
+      link: ''
+    }))
+  } else {
+    throw new Error('Failed to fetch levels: Response is not an array.')
```
Comment (fromgonie) [inline frontend/src/stores/levels/actions.ts:15]: You don`t need this error

File: frontend/src/views/Viewlevelstable.vue [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+  </div>
+</template>
+
+<script setup lang="ts">
+import { ref, onMounted } from 'vue'
+import { useRouter, useRoute } from 'vue-router'
+import { ElTable, ElTableColumn, ElButton } from 'element-plus'
+import { useLevelsStore } from '@/stores/levels'
+import { fetchLevelsByRoleId } from '@/stores/levels/actions'
+import { getLevels } from '@/stores/levels/getters'
```
Comment (fromgonie) [inline frontend/src/views/Viewlevelstable.vue:22]: Line 21 and 22 you are using store functions directly, and this is not supposed to be done. Use levelsStore.functionYouWantToUse

File: frontend/src/views/Viewlevelstable.vue [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+const levelsStore = useLevelsStore()
+const levels = ref([])
+
+const roleId = route.params.roleId?.toString() || ''
+
+const fetchLevels = async () => {
+  try {
+    await fetchLevelsByRoleId(roleId)
+    const fetchedLevels = getLevels()
+    if (Array.isArray(fetchedLevels)) {
```
Comment (fromgonie) [inline frontend/src/views/Viewlevelstable.vue:35]: You don`t need this check here

File: frontend/src/router/index.ts [hunk 28]
```diff
@@ -31,6 +32,11 @@ const router = createRouter({
       path: '/level-view',
       component: LevelView
     },
+    {
+      path: '/level-table/:roleId',
+      component: ViewLevelsTable,
+      props: true // Enable passing route params as props directly
```
Comment (vitorlofonseca) [inline frontend/src/router/index.ts:38]:
```suggestion
      props: true

File: frontend/src/stores/levels/actions.ts [hunk 29]
```diff
@@ -0,0 +1,34 @@
+import { levels } from './state'
+import { get } from '@/services/http'
+
+export interface Level {
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:4]: This interface already exists under domain folder. Please, get rid of it, and use the interface of domain (if it doesn't exist, create like this)

File: frontend/src/stores/levels/actions.ts [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+  name: string
+  minCoefficient: number
+  maxCoefficient: number
+  link: string
+}
+
+export async function fetchLevelsByRoleId(roleId: string): Promise<Level[]> {
+  const data: Level[] = await get(`/levels?role_id=${roleId}`)
+
+  const levelsData: Level[] = data.map((level) => ({
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:15]: You don't need this...

File: frontend/src/stores/levels/actions.ts [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+    maxCoefficient: 0,
+    link: ''
+  }))
+
+  levels.value = levelsData
+
+  return levelsData
+}
+
+fetchLevelsByRoleId('someRoleId')
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:28]: Can you remove this call?

File: frontend/src/stores/levels/actions.ts [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+
+  const levelsData: Level[] = data.map((level) => ({
+    id: level.id,
+    name: level.name,
+    minCoefficient: 0,
+    maxCoefficient: 0,
+    link: ''
+  }))
+
+  levels.value = levelsData
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:23]: You don't need a state for levels. You can put the levels of a role inside the role state. And then, when you call this method, you should verify if the levels of the role already exists (in case yes, you just return)

File: frontend/src/stores/levels/state.ts [hunk 33]
```diff
@@ -0,0 +1,6 @@
+import type { Level } from '@/domain/Level'
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/state.ts:1]: Get rid of this file, please

File: frontend/src/stores/levels/index.ts
Comment (vitorlofonseca) [inline frontend/src/stores/levels/index.ts:1]: Get rid of this file, please

File: frontend/src/stores/roles/state.ts [hunk 34]
```diff
@@ -1,6 +1,11 @@
+import { ref, type Ref } from 'vue'
 import type { Role } from '@/domain/Role'
-import { type Ref, ref } from 'vue'
+import type { Level } from '@/domain/Level'
 
-const roles: Ref<Role[]> = ref([])
+export interface RoleWithLevels extends Role {
```
Comment (vitorlofonseca) [inline frontend/src/stores/roles/state.ts:5]: You don't need to do this, you already have a levels parameter inside Role object

image

File: frontend/src/views/ViewLevelsTable.vue [hunk 35]
```diff
@@ -0,0 +1,67 @@
...
+const levels = ref([])
+
+const roleId = route.params.roleId?.toString() || ''
+
+const fetchLevels = async () => {
+  try {
+    await levelsStore.fetchLevelsByRoleId(roleId)
+    levels.value = levelsStore.getLevels()
+  } catch (error) {
+    console.error('Failed to fetch levels:', error)
```
Comment (vitorlofonseca) [inline frontend/src/views/ViewLevelsTable.vue:35]: No console messages! Use the messages to the user

File: frontend/src/stores/levels/getters.ts
Comment (vitorlofonseca) [inline frontend/src/stores/levels/getters.ts:1]: Please, get rid of this file

File: frontend/src/stores/levels/actions.ts [hunk 36]
```diff
@@ -0,0 +1,47 @@
+import { get, post, put, deleteRequest } from '@/services/http'
+import { levels } from './state'
+import type { Level } from '@/domain/Level'
+
+// Function to sort levels based on minCoefficient and maxCoefficient
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:5]: This should be done in the backend

File: frontend/src/stores/levels/actions.ts [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+function sortLevels(levels: Level[]): Level[] {
+  return levels.slice().sort((a, b) => {
+    return a.minCoefficient - b.minCoefficient || a.maxCoefficient - b.maxCoefficient
+  })
+}
+
+async function fetchLevels(): Promise<void> {
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:12]: This function is useless, you'll fetch the levels of just one level, not all levels

File: frontend/src/stores/levels/actions.ts [hunk 38]
```diff
@@ -0,0 +1,47 @@
...
+  const fetchedLevels = await get<Level[]>('/level')
+  setLevels(fetchedLevels)
+}
+
+function setLevels(newLevels: Level[]): void {
+  // Sort fetched levels before setting them
+  levels.value = sortLevels(newLevels)
+}
+
+function removeLevel(levelId: number) {
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:26]: Your ticket is about **list the levels of a role**. Why have you created functions to remove, save and edit???

File: frontend/src/stores/levels/actions.ts [hunk 39]
```diff
@@ -0,0 +1,47 @@
+import { get, post, put, deleteRequest } from '@/services/http'
+import { levels } from './state'
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:2]: You don't need this state. You have to set your levels to the role you are passing

1. Get the levels in the backend, of a specific role
2. Put those levels in the role object, in the role state. For this, you'll need to import the role state here in this file

File: frontend/src/views/ViewLevelsTable.vue [hunk 40]
```diff
@@ -0,0 +1,111 @@
...
+    <div class="TableContainer">
+      <h1>{{ role?.name }}</h1>
+      <h2>Attached Levels</h2>
+      <br />
+      <ElTable :data="sortedLevels" style="width: 100%">
+        <ElTableColumn prop="name" label="Levels" />
+        <ElTableColumn fixed="right" label="Actions" width="150">
+          <template #default="{ row }">
+            <div>
+              <ElButton @click="onDeleteRow(row)" type="text" size="small">Delete</ElButton>
```
Comment (vitorlofonseca) [inline frontend/src/views/ViewLevelsTable.vue:12]: Remove EVERYTHING not related to the level table here

File: frontend/src/stores/levels/actions.ts [hunk 41]
```diff
@@ -0,0 +1,16 @@
...
+
+async function fetchLevelsByRoleId(roleId: number): Promise<void> {
+  const role = roles.value.find((role) => role.id === roleId)
+
+  if (role && !role.levels) {
+    const sortedLevels = await get<Level[]>(`/levels/sorted/${roleId}`)
+    return role.levels
+
+    role.levels = sortedLevels
+  }
```
Comment (vitorlofonseca) [inline frontend/src/stores/levels/actions.ts:13]: You have to return the role.levels here

File: backend/src/main/java/saci/infrastructure/LevelRepository.java [hunk 42]
```diff
@@ -10,10 +12,16 @@
...
     @Query(
             "SELECT COUNT(*) FROM Level l WHERE l.roleId = :roleId AND (l.minCoefficient < :maxCoefficient AND l.maxCoefficient > :minCoefficient)")
     int overlappingLevelsCounter(
             @Param("roleId") Long roleId,
             @Param("minCoefficient") Integer minCoefficient,
             @Param("maxCoefficient") Integer maxCoefficient);
+
+    @Query(
+            "SELECT l FROM Level l WHERE l.roleId = :roleId ORDER BY l.minCoefficient ASC, l.maxCoefficient ASC")
+    List<Level> findByRoleIdOrderByMinCoefficientAscMaxCoefficientAsc(@Param("roleId") Long roleId);
```
Comment (vascolusitan) [inline backend/src/main/java/saci/infrastructure/LevelRepository.java:26]: Change this method name to a simpler one please, it's hard to understand easily why you do this.

**Review by vitorlofonseca — APPROVED** [review]

**Review by fromgonie — APPROVED** [review]
